### ISSUE TITLE
Calling AbortController abort causes mutation promise to never resolve 

### ISSUE THREAD

**Issue by destin-estrela** [apollographql/apollo-client#10640]
### Issue Description

I am using `AbortController` to cancel mutations that are taking too long to receive a response in my application. However, calling abort causes the promise to never complete. Not even the error or finally of a try-catch-finally block will execute. 

Removing this the below part of the source code gives me exactly my intended behavior: catching and reporting an error if the call is aborted.

```ts
// apollo-client/src/link/http/parseAndCheckHttpResponse.ts:133

export function handleError(err: any, observer: Observer<any>) {
  if (err.name === "AbortError") return; // removing this fixes everything!
```

Since the above if statement feels very intentional, I might have a misunderstanding of how the behavior should be. If there is no bug, please help me clarify how I should handle this situation. 

### Link to Reproduction

https://codesandbox.io/s/quirky-golick-o14tos?file=/src/index.js

### Reproduction Steps

Click create item (without bug). Observe in the console that the promise will eventually resolve and that the finally executes.

Now click create item (with bug). Observe in the console that abort is called, and that the promise never resolves. Observe that no errors are reported and that the `console.log` in the `finally` are not executed.

**Comment by bignimbus** [comment]
Hi @destin-estrela 👋🏻  I forked your reproduction [here](https://codesandbox.io/s/wizardly-cannon-so2jgs?file=/src/index.js) and added a `fetchOptions.signal` to your `useMutation` options argument and removed the `context` argument provided to the `addTodo` mutation function.  I believe the context isn't doing what you expect it to do in this instance.  Does that address the issue you're seeing?  Here's the relevant snippet:

```js
function AddTodo() {
  const controller = new AbortController();
  const [addTodo] = useMutation(ADD_TODO, {
    fetchOptions: {
      signal: controller.signal
    }
  });

  const handleSubmit = async (shouldAbort) => {
    if (shouldAbort) {
      setTimeout(() => {
        controller.abort();
        console.log("Aborted");
      }, 1);
    }

    try {
      console.log("About to add todo");
      await addTodo({
        variables: { description: "X" }
      });
      console.log("Promise resolved!");
    } catch (error) {
      console.error(error);
    } finally {
      console.log("Finally ran!");
    }
  };
  // ...
}
```

**Comment by destin-estrela** [comment]
Hello, thank you for your reply!

 I did a lot of experimenting in both my own project and your forked example. It looks like what you provided doesn't work. If you inspect the request in the dev tools, you can observe that it is no longer aborted like the one I supplied.

I also could find "fetchOptions" an acceptable type within the useMutation hook when using @apollo/client with typescript.

**Comment by phryneas** [comment]
Yeah, it should be 
```ts
  const [addTodo] = useMutation(ADD_TODO, {
    context: {
      fetchOptions: {
        signal: controller.signal,
      },
    },
  });
```
I can reproduce the issue over in https://codesandbox.io/p/sandbox/red-snow-n17dci

**Comment by labithiotis** [comment]
I have a workaround that should work for anyone needing to abort a mutation, it will create a new abortController for each new request and when aborted will set the request loading to false. It may not be perfect but should work for now.

```tsx
import { useCallback, useRef, useState } from 'react';
import { ApolloCache, DefaultContext, MutationOptions, OperationVariables, useApolloClient } from '@apollo/client';

export function useAbortableMutation<
  TData = any,
  TVariables extends OperationVariables = OperationVariables,
  TContext extends Record<string, any> = DefaultContext,
  TCache extends ApolloCache<any> = ApolloCache<any>,
>(
  mutation: any,
): [
  (options: Omit<MutationOptions<TData, TVariables, TContext>, 'mutation'>) => Promise<void>,
  { data?: TData | null; error?: Error; loading: boolean },
  () => void,
] {
  const client = useApolloClient();
  const controller = useRef(new window.AbortController());
  const [data, setData] = useState<TData | null>();
  const [error, setError] = useState<Error>();
  const [loading, setLoading] = useState(false);

  const request = useCallback(
    async (options: Omit<MutationOptions<TData, TVariables, TContext>, 'mutation'>) => {
      try {
        setLoading(true);
        controller.current = new window.AbortController();
        const res = await client.mutate<TData, TVariables, TContext, TCache>({
          ...options,
          mutation,
          context: { ...options.context, fetchOptions: { signal: controller.current.signal } } as unknown as TContext,
          errorPolicy: 'all',
        });
        setData(res.data);
      } catch (error) {
        if (error instanceof Error) {
          setError(error);
        }
      } finally {
        setLoading(false);
      }
    },
    [client, mutation],
  );

  const abort = useCallback(() => {
    if (!controller.current.signal.aborted) {
      controller.current.abort();
      setLoading(false);
    }
  }, [controller]);

  return [request, { data, error, loading }, abort];
}
```

```tsx
export function() {
  const  [doSomething, { loading, error, data }, abort] = useAbortableMutation<Variables, Data>(gql`mutation DoSomething() { doSomething() }`);  
}
```

**Comment by jerelmiller** [comment]
Fixed by #11058. This change will go out with our `rc` release of v3.8.0, which will happen very soon!
